ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java

```diff
@@ -197,27 +136,46 @@ public void onError(Throwable t) {
                     t instanceof NullPointerException ? "NullPointerException" : t.getMessage(),
                     t
                 );
-            response(tensorOutputs);
+            response();
         }
 
         @Override
         public void onComplete() {
-            response(tensorOutputs);
+            response();
         }
     }
 
-    private void response(Map<Integer, ModelTensors> tensors) {
-        processResponse(statusCode, responseBody.toString(), parameters, tensorOutputs);
-        executionContext.getCountDownLatch().countDown();
-        // when countdown's count equals to 0 means all responses are received.
-        if (executionContext.getCountDownLatch().getCount() == 0) {
-            if (executionContext.getExceptionHolder().get() != null) {
-                actionListener.onFailure(executionContext.getExceptionHolder().get());
-                return;
+    private void response() {
+        String body = responseBody.toString();
+        if (Strings.isBlank(body)) {
+            log.error("Remote model response body is empty!");
+            actionListener.onFailure(new OpenSearchStatusException("No response from model", RestStatus.BAD_REQUEST));
+            return;
+        }
+
+        if (statusCode < HttpStatus.SC_OK || statusCode > HttpStatus.SC_MULTIPLE_CHOICES) {
+            if (containsThrottlingException) {
```

Comment:
Do we really need to make retry action specific for throttling? Can we make it general for all exceptions?

Reply:
Not all exceptions should be retried, e.g. GPU OOM exception, malformed parameters etc. Retry for these exceptions will only increase the load of remote endpoint, but won't get it run successfully.
For now we only find the throttling exception is suitable for retry. In the future we can extend it to more cases based on our new findings